Sources/SwiftParser/StringLiterals.swift

```diff
@@ -0,0 +1,126 @@
+
+extension Parser {
+  /// Consumes a raw string delimiter that has the same number of `#` as `openDelimiter`.
+  private mutating func parseStringDelimiter(openDelimiter: RawTokenSyntax?) -> (unexpectedBeforeCheckedDelimiter: RawUnexpectedNodesSyntax?, checkedDelimiter: RawTokenSyntax?) {
+    // Check for leadingTriviaText == "" so we don't consume the leading raw
+    // string delimiter of an upcoming string literal, e.g. in
+    // ```
+    // "normal literal"
+    // #"raw literal"#
+    // ```
+    let delimiter: RawTokenSyntax?
+    if self.at(.rawStringDelimiter) && self.currentToken.leadingTriviaText == "" {
+      delimiter = self.consumeAnyToken()
+    } else {
+      delimiter = nil
+    }
+
+    switch (openDelimiter, delimiter) {
+    case (nil, nil):
+      return (nil, nil)
+    case (let open?, nil):
+      return (nil, missingToken(.rawStringDelimiter, text: open.tokenText))
+    case (nil, .some):
+      return (RawUnexpectedNodesSyntax([delimiter], arena: self.arena), nil)
+    case (let open?, let close?):
+      if open.tokenText == close.tokenText {
+        return (nil, close)
+      } else {
+        return (RawUnexpectedNodesSyntax([delimiter], arena: self.arena), missingToken(.rawStringDelimiter, text: open.tokenText))
+      }
+    }
+  }
+
+  /// Parse a string literal expression.
+  @_spi(RawSyntax)
+  public mutating func parseStringLiteral() -> RawStringLiteralExprSyntax {
+    /// Parse opening raw string delimiter if exist.
+    let openDelimiter = self.consume(if: .rawStringDelimiter)
+
+    /// Parse open quote.
+    let (unexpectedBeforeOpenQuote, openQuote) = self.expectAny([.stringQuote, .multilineStringQuote, .singleQuote], default: .stringQuote)
+
+    /// Parse segments.
+    var segments: [RawStringLiteralSegmentsSyntax.Element] = []
+    var loopProgress = LoopProgressCondition()
+    while loopProgress.evaluate(self.currentToken) {
+      if let stringSegment = self.consume(if: .stringSegment) {
+        segments.append(.stringSegment(RawStringSegmentSyntax(content: stringSegment, arena: self.arena)))
+      } else if let backslash = self.consume(if: .backslash) {
+        let (unexpectedBeforeDelimiter, delimiter) = self.parseStringDelimiter(openDelimiter: openDelimiter)
+        let (unexpectedBeforeLeftParen, leftParen) = self.expect(.leftParen)
+        let expressions = RawTupleExprElementListSyntax(elements: self.parseArgumentListElements(pattern: .none), arena: self.arena)
+
+        // For recovery, eat anything up to the next token that either starts a new string segment or terminates the string.
+        // This allows us to skip over extraneous identifiers etc. in an unterminated string interpolation.
+        var unexpectedBeforeRightParen: [RawTokenSyntax] = []
+        var unexpectedProgress = LoopProgressCondition()
+        while !self.at(any: [.rightParen, .stringSegment, .backslash, openQuote.tokenKind, .eof]) && unexpectedProgress.evaluate(self.currentToken) {
```

Comment:
In single line string literals, this should stop at newline IMO.

Reply:
The lexer automatically produces a `stringSegment` at the end of a single-line string literal, even if it is not terminated (see the new test `testNewlineInInterpolationOfSingleLineString` in LexerTest.swift), so we will stop at that.